### DISCUSSION TITLE
Questions about how the stack is allocated when using arduino-pico with platformio

### DISCUSSION THREAD

**Discussion by JABCanford** [earlephilhower/arduino-pico#2634] [Q&A]
I've got a few questions about how arduino-pico allocates the stack. I'm working on a project using arduino-pico in platformio, with a custom board based on the Wiznet W5500-EVB, but with some extra hardware (relays, LCD, etc).

I recently had a crash inducing problem with my program which went away when I moved a 100 byte array from being a local variable to a global variable, leading me to try and find out if the stack being full was the cause of my issue. 

I tried adding "bool core1_separate_stack = true;" to my program, as I am using both cores, to try and increase the stack size for core 0, as documented by [https://arduino-pico.readthedocs.io/en/latest/multicore.html
/](https://arduino-pico.readthedocs.io/en/latest/multicore.html). I ran platformio's memory inspector, to see if the stack was being made bigger by adding this, ie core 0 getting all 8k of the scratch memory and core 1 getting 8k from the main SRAM. The inspector seems to show the stack for core 0 occupying the lower half of scratch_y, and the stack for core 1 occupying the lower half of scratch_x. I'm not sure if either the core1_separate_stack flag isn't working, or if platformio's inspector isn't spotting the changes that the flag is causing.

![pasted_image](https://github.com/user-attachments/assets/86c0e50e-4883-47e1-9f4c-e4793384a6e1)

Is there any significance to the word "dummy" in the name ".stack_dummy", beyond that the compiler doesn't know how much space will be used at runtime?
Is the 2k size of ".stack_dummy" a limit, or can the stack use the whole 4k of the scratch by default (in multicore)?
Please forgive anything I'm getting wrong here, this is the first time I've written a large enough arduino-pico program to encounter issues with the stack.

When trying to solve the crashes I was previously having, I wanted to see how much was on the stack at runtime, so wanted to use the "inline int getFreeStack()" function added by a commit from pull #1816 (https://github.com/earlephilhower/arduino-pico/pull/1816), but I wasn't able to get this to work in my program, despite having other functions from the rp2040 class such as "idleOtherCore()" work fine. Is this function intended to be user accessible?

As suggested in https://community.platformio.org/t/a-nice-way-to-see-exact-memory-usage/32118, I added -Wl,--print-memory-usage to build_flags: in platformio.ini. This gave me extra output when building:

```
Memory region         Used Size  Region Size  %age Used
           FLASH:      183256 B      2044 KB      8.76%
             RAM:       18532 B       256 KB      7.07%
       SCRATCH_X:          2 KB         4 KB     50.00%
       SCRATCH_Y:          0 GB         4 KB      0.00%
```

This also doesn't seem right, as I would have thought that scratch_y should show some usage for the stack for core 0, given that scratch_x is showing usage?

Thank you in advance for any explanations.

**Comment by maxgerhardt** [comment]
This should answer some questions:

https://github.com/earlephilhower/arduino-pico/blob/9cea4708c8eaf3ff7cf2f9dc1ff73dafc00bd91e/lib/rp2040/memmap_default.ld#L256-L272

https://github.com/raspberrypi/pico-sdk/blob/95ea6acad131124694cda1c162c52cd30e0aece0/src/rp2_common/pico_crt0/crt0.S#L533-L537

https://github.com/raspberrypi/pico-sdk/blob/95ea6acad131124694cda1c162c52cd30e0aece0/src/rp2040/pico_platform/include/pico/platform.h#L29-L32

https://github.com/raspberrypi/pico-sdk/blob/95ea6acad131124694cda1c162c52cd30e0aece0/src/rp2_common/pico_clib_interface/newlib_interface.c#L180-L184

How did you try to use `RP2040.getFreeStack()`? It should work.

**Reply by JABCanford** [reply]
So far I've just been trying to have `getFreeStack()` return an int. I get compilation errors if I include it.

When compiling with 
```c++
int checkStack(){
  int freeStack = RP2040.getFreeStack();
  return freeStack;
}
``` 
I get the error
```
src/main.cpp: In function 'int checkStack()':
src/main.cpp:6422:25: error: expected primary-expression before '.' token
 6422 |   int freeStack = RP2040.getFreeStack();
      |                         ^
```


And if I change `RP2040` to `rp2040` so my function is
```c++
int checkStack(){
  int freeStack = rp2040.getFreeStack();
  return freeStack;
}
``` 
I get the error
```
src/main.cpp: In function 'int checkStack()':
src/main.cpp:6422:26: error: 'class RP2040' has no member named 'getFreeStack'; did you mean 'getFreeHeap'?
 6422 |   int freeStack = rp2040.getFreeStack();
      |                          ^~~~~~~~~~~~
      |                          getFreeHeap
```

**Reply by maxgerhardt** [reply]
Are you sure your platform integration is up-to-date? You might be experiencing bugs which have been solved since some eternity. This code works for me.

![grafik](https://github.com/user-attachments/assets/be8fb8db-1c4f-40ee-9224-3b578efb0457)

To update: Open a [PlatformIO core CLI](https://docs.platformio.org/en/latest/core/index.html) -> 
```
pio pkg update -g -p "https://github.com/maxgerhardt/platform-raspberrypi.git"
```
(or `-p raspberrypi` if this is your only installed raspberrypi platform)

**Reply by JABCanford** [reply]
Thanks, I hadn't realised that my copy of arduino-pico was so outdated. Updating it and restarting VS Code has fixed that issue and `rp2040.getFreeStack()` now produces no compiler errors. 

However, I now have an error for the declaration of `bool core1_separate_stack = true;`
```
In file included from /home/jb/.platformio/packages/framework-arduinopico/cores/rp2040/Arduino.h:129,
                 from src/main.cpp:56:
/home/jb/.platformio/packages/framework-arduinopico/cores/rp2040/RP2040Support.h:173:17: error: conflicting declaration of 'bool core1_separate_stack' with 'C' linkage
  173 | extern "C" bool core1_separate_stack;
      |                 ^~~~~~~~~~~~~~~~~~~~
src/main.cpp:50:6: note: previous declaration with 'C++' linkage
   50 | bool core1_separate_stack = true; //increased stack size, see https://arduino-pico.readthedocs.io/en/latest/multicore.html
      |      ^~~~~~~~~~~~~~~~~~~~
``` 
I think I'm using it the way intended by [the documentation](https://arduino-pico.readthedocs.io/en/latest/multicore.html)?

**Reply by JABCanford** [reply]
I've found that adding `extern "C"` to the declaration in my code eliminates the above error.

Using `getFreeStack()`, I get 4096 bytes more when `extern "C" bool core1_separate_stack = true;` is included than when it isn't, so it seems to be working.

However, the memory inspector in platformio still shows the stacks in the same locations in memory (0x20041000 and 0x20040000), so either platformio's inspector is wrong, or both stacks are still in the scratch RAM and getting 4k each.
